### PR TITLE
Use middleware to synchronize store to history

### PR DESCRIPTION
**Pull request by taion**
Per https://github.com/rackt/redux-simple-router/pull/129#issuecomment-167392040.

Fixes https://github.com/rackt/redux-simple-router/issues/108.

The idea is basically - keep unidirectional data flow for updates, most of the time. This works in most cases, but it breaks replay, since the actions in some sense aren't pure. That's fine, though - we can restrict the store observer to _only_ handle the replay case (and potentially not even use it otherwise).

### CODE REVIEW

**Comment by taion** [comment]
This actually fixes https://github.com/rackt/redux-simple-router/issues/140 as well.

**Comment by taion** [comment]
I believe this fixes https://github.com/rackt/redux-simple-router/issues/122, too.

**Comment by taion** [comment]
Probably also https://github.com/rackt/redux-simple-router/issues/113, since I think using `location.key` avoids the edge cases you get from an incrementing `changeId`.

**Comment by taion** [comment]
Note that you cannot in general directly update the store based on just the action - it's possible for e.g. a `listenBefore` hook to block the transition, and that would then take the store out-of-sync with the browser.

**Comment by tomatau** [comment]
Just tried it out and it does fix https://github.com/rackt/redux-simple-router/issues/108 with a nice implementation. Feels more suitable to redux as a simple middleware.

I think it would be worth updating usage and example in this PR for clarity.

**Comment by ezekielchentnik** [comment]
@taion I believe location.key is not unique, it's random - different every time .push/replace are called, regardless if same path, state.  not sure if that matters...

**Comment by taion** [comment]
I know exactly what `location.key` is, thanks.

**Comment by taion** [comment]
@tomatau I see you beat me to it. The one wart is that you need `middleware.syncHistoryToStore` per https://github.com/rackt/redux-simple-router/pull/141/files#diff-413f2628e9c93c46cec6db202ddb8ea4R159 if you want the Redux DevTools to work (but in normal operation that bit is a no-op).

Wanted to get some feedback from @jlongster and @kjbekkelund as to whether this approach seems workable. I don't use Redux at all myself, but it seems like preserving unidirectional data flow via a transducer just gets rid of all the cycle management headaches (though it requires that extra line of boilerplate for DevTool support).

This is halfway a PoC just to illustrate what's involved though. Ideally, for a breaking change, it'd be best to make the user-dispatched actions take location descriptors (and correspondingly probably just have a different `type`), while having the reducer operate on actual locations from `history`.

**Comment by kjbekkelund** [comment]
@taion Very interesting! I'm quite busy until early January, but hopefully @jlongster has got some time to review and test. Otherwise I'll test as soon as I've got some spare hours :)

**Comment by tomatau** [comment]
> The one wart is that you need middleware.syncHistoryToStore per https://github.com/rackt/redux-simple-router/pull/141/files#diff-413f2628e9c93c46cec6db202ddb8ea4R159 if you want the Redux DevTools to work (but in normal operation that bit is a no-op).

Testing this out seems buggy as it's dispatching new updatePath actions when rolling back history... which is very bad and counter-intuitive for debugging. Not adding this line seems like a nicer experience, to just go without reversing through the path history from DevTools rollbacks.

I appreciate that this is mainly a PoC, but I'm very keen to see one of these new implementations get merged and fix up https://github.com/rackt/redux-simple-router/issues/108 ASAP :) So put the word out for any work to get movement and I'm happy to contribute.

**Comment by taion** [comment]
It's tricky. Suppose you want to reverse the path history.

This necessarily has to have a side effect. I think the right way to model this on the `history` side is with a `push`, but that will create a new `location` with a new `location.key`.

If I don't dispatch the extra "update path" action, then the `key` for the `location` would be out of sync between `history` and `store`.

It's possible to directly use `history.transitionTo` and keep the same `key`, but that means we'd want to preserve the action involved, and it'd seem odd to me to issue a `REPLACE` or a `POP` to the `history` rather than a `PUSH` here.

**Comment by taion** [comment]
Well, maybe using `history.transitionTo` would be all right.

**Comment by jlongster** [comment]
I've been resisting to forcing a middleware on the user, as I think a middleware should only be for converting one action into multiple actions, but I haven't had time to look closely at this. I will do so this week though.

Also, long-term you and other people who work on react-router will probably be involved in maintaining this so you have a greater say in how this evolves than other PRs in my opinion.

**Comment by taion** [comment]
I think the problem with round-tripping this through the store is that it breaks unidirectional data flow. If you do this before it hits the store in the first place, and only use the store for restoring location on replay, you completely exclude issues like #108; otherwise it just feels like a bunch of work to avoid cascading updates, which Flux-like architectures should inherently exclude.

**Comment by jlongster** [comment]
Yeah, there are benefits to it, but I'd say that #129 is still not that much work and it avoids other problems that this may introduce (there are so many little edge cases with the devtools), without the need for the user to install a middleware. I'm open to it! Just need to look at it more.

**Comment by taion** [comment]
Let me flesh this out with the changes we mentioned earlier.

**Comment by taion** [comment]
Updated the PR. This now closes https://github.com/rackt/redux-simple-router/issues/95 as well. It's a more significant API change, though.

**Comment by taion** [comment]
@tomatau 

This should fix your https://github.com/rackt/redux-simple-router/pull/141#issuecomment-167618037, except when completely clearing all the actions.

The idea is that this adds the same asymmetry between providing a location descriptor to the action creators, while populating the store with fully fleshed-out location objects.

**Comment by jlongster** [comment]
@taion I've always wanted to avoid forcing the user to install a middleware. However, I don't really care that much. Really, the two core ideas that need to exist for me to be happy are: only dumb JS objects stored in the app state, and the user uses react-router's API directly in routing components (and wherever else they. As long as that's the case, I don't really care how it works.

You have a much better understanding of how more complex stuff like async routing works also.

If we pursue this path, are you going to continually be involved here? I foresee myself being less involved over time, as I was really hoping to kick off this new direction (and I'm happy it succeeded). I'd love it if you and other react-router folks eventually took over maintenance, but I don't know how burdened you all are with that already.

File: src/index.js [hunk 1]
```diff
@@ -1,149 +1,107 @@
...
+  })
 }
 
-export function replacePath(path, state, { avoidRouterUpdate = false } = {}) {
+export const push = transition('push')
+export const replace = transition('replace')
+
+// TODO: Add go, goBack, goForward.
+
+function updateLocation(location) {
```
Comment (jlongster) [inline src/index.js]: I'm assuming this `location` object is now the fully parsed location? Can we make it so you can also pass just a string, and if you do that we just convert it into `{ pathname: location }`?

Comment (jlongster) [inline src/index.js]: I know that makes the API slightly asymmetrical (the state would be different than what you passed in), but now that we can always store the same shape of object in the state I think it's ok.

Comment (taion) [inline src/index.js]: It's already supported. See e.g. https://github.com/rackt/redux-simple-router/pull/141/files#diff-413f2628e9c93c46cec6db202ddb8ea4R338.

Comment (taion) [inline src/index.js]: Oops - misread. I'm actually not exporting `updateLocation` at all. The `transition` actions do take strings, though.

File: src/index.js [hunk 2]
```diff
@@ -1,149 +1,107 @@
...
-  }
+function transition(method) {
+  return arg => ({
+    type: TRANSITION,
+    method, arg
+  })
 }
 
-export function replacePath(path, state, { avoidRouterUpdate = false } = {}) {
+export const push = transition('push')
```
Comment (jlongster) [inline src/index.js]: `push` and `replace` are *very* vague names when seen outside of react-router. Would you be ok keeping the Path suffix? I think it generally makes it clearer.

File: src/index.js (hunk 1 above)
Comment (jlongster) [inline src/index.js]: oh, great, looks like that's a natively-supported thing in react-router.

File: src/index.js (hunk 2 above)
Comment (taion) [inline src/index.js]: I agree that `push` and `replace` are awful names. Maybe `pushLocation` and `replaceLocation`? I thought `pushPath` was a bit inaccurate, since this would let you push `query`, &c. as well. I also didn't want to figure out how to then name `go`, `goBack`, `goForward` – would we suffix those with `location` as well?

File: src/index.js (hunk 1 above)
Comment (taion) [inline src/index.js]: Yes – from the `TRANSITION` actions. This is just for receiving the update from `history` for the reducer.

**Comment by taion** [comment]
This doesn't touch on async routing at all – and it can't, without a much larger re-write. The routing data is only exposed to the children of `<Router>` right now. By contrast, as you've discussed elsewhere, the library as-is only integrates into `history`. It's difficult; I can see some ways to make it work, but since I don't actually use Redux, I'd have no clue whether the API would be satisfactory to users or not.

As for my time involvement, I don't know – I don't use Redux, so this is just my vague impressionistic idea of how a Redux integration might look from what I've seen on paper.

Maybe to break down the pieces here a little bit, though, there are a few things here worth highlighting:

- I feel like the primary purpose of an integration like this is to sync history and store in normal use. Intercepting everything in a middleware gives you that, because the data flow is just either action -> history -> store, or history -> store; you don't have to deal with both an action -> store -> history path and a separate history -> store path.
  - Correspondingly, the dev tools integration to _cause_ side effects on modifying action history is secondary; if the user doesn't want this, he or she doesn't have to use `middleware.syncHistoryToStore` at all, so it's the same amount of API work – just a matter of installing the middleware instead of syncing the `history` to the `store`
- With the new history APIs, the natural point of integration is having the user specify a location descriptor, but putting location objects in the store. Among other things, this is how all the `history` APIs will work, and it can allow e.g. supporting named routes in a very generic way (`history.push({ name: 'foo', params })`) and have things _just work_ for a user of this library. This means, though, that the "raw" user actions probably shouldn't hit the reducer without going through `history` first, since you want locations at the reducer, not location descriptors.
- Using `changeId` seems quite fragile – you're probably better off using `location.key` instead, and using `history.transitionTo` to restore the key after a dev tools action.

**Comment by taion** [comment]
To elaborate, with location descriptors, you really want to separate the location descriptor the user provides from the location that goes into the store. That means the current approach of feeding the user-supplied action directly into the reducer is unlikely to work well.

You could e.g. have a "last location descriptor action" reducer that a store subscriber listens to, then uses to update `history`, and you'd still have the same architectural design where all updates to the actual reducer for the locations comes from the `history` (necessarily so), but I just think a middleware is cleaner (really, more unidirectional) here.

**Comment by jlongster** [comment]
Yeah, I think you've convinced me. The big win is that what is stored in the state is a consistent location object coming from history, no matter what the user called the action creator with. This also opens up the path to supporting routing params if we want to do that.

Note that #129 does not use `changeId`, it's a different implementation and you should look at that for the "latest" ideas.

**Comment by taion** [comment]
Just updated the PR again to fix an edge case.

What do you think of the middleware swallowing the `TRANSITION` action? I set it up that way to avoid having to change the `'only triggers store once when updating path via store'` test, but maybe that's not so correct - the default reducer already ignores the `TRANSITION` action anyway.

**Comment by jlongster** [comment]
What happens in that test differently now? Why is the store updated more than once?

Reading your comment: "We could also use the same type in both places" How could we use the same action type, wouldn't that be an infinite loop because it keeps triggering the middleware?

**Comment by taion** [comment]
1. If I don't swallow the `TRANSITION` action, I see the store subscriber firing on both the `TRANSITION` action (for a no-op non-change) and the `UPDATE_LOCATION` action (that does cause the store to update)
2. I probably should have added a test case - just tracing through the logic on paper, if you reset all the actions while on the initial location, I think my implementation per https://github.com/rackt/redux-simple-router/commit/243235eb7c1c9c5c79ab22e2e1eddc88123357d2 would have left the store empty, since we would have had `location.key === currentKey`. So in the special case where the store is cleared, we need to make `history` cause an `UPDATE_LOCATION` action to be dispatched regardless.
3. This is probably really messy, but you could do something along the lines of e.g. https://github.com/rackt/redux-simple-router/commit/6fc6ab7a5fd29cf8cb7ab0f39a528a10cdc30d5f#diff-1fdf421c05c1140f6d71444ea2b27638R71, where actions that don't have e.g. a `location` key just get swallowed and sent to `history` instead (but ones with a `location` key get passed through). This is probably as horrible/hacky as it sounds, right?

**Comment by jlongster** [comment]
> If I don't swallow the TRANSITION action, I see the store subscriber firing on both the TRANSITION action (for a no-op non-change) and the UPDATE_LOCATION action (that does cause the store to update)

Ah, right.

> I probably should have added a test case - just tracing through the logic on paper, if you reset all the actions while on the initial location, I think my implementation per 243235e would have left the store empty, since we would have had location.key === currentKey. So in the special case where the store is cleared, we need to make history cause an UPDATE_LOCATION action to be dispatched regardless.

Yeah, clearing all actions is a weird edge case to support from the redux devtools. Mainly because the initial routing state is actually wrong and it requires one action during init to be in sync with the current location.

It sounds like you've been able to figure it out though.

> This is probably really messy, but you could do something along the lines of e.g. 6fc6ab7#diff-1fdf421c05c1140f6d71444ea2b27638R71, where actions that don't have e.g. a location key just get swallowed and sent to history instead (but ones with a location key get passed through). This is probably as horrible/hacky as it sounds, right?

Yeah, I think I like having two action types better. It's clearer and easier to debug.

Redux devtools won't see any of the `TRANSITION` actions, but only `UPDATE_LOCATION`, right?

File: src/index.js (hunk 2 above)
Comment (jlongster) [inline src/index.js]: Hm that's a good point. `goPath` doesn't really sound great. We could just rename `push` and `replace` and leave the other be, but if you think that's confusing we can just stick with these simple names and users can rename them when importing.

**Comment by jlongster** [comment]
I'll actually use this tomorrow (or, worst case later this week), but if it seems to work well I'm tempted to just merge it. We need to settle on something, and this sounds like the most promising solution so far.

File: src/index.js (hunk 2 above)
Comment (taion) [inline src/index.js]: I don't know the naming conventions for Redux action creators at all. Is there an option to namescope these under e.g. `HistoryActionCreators` or something, like it's done with the devtools action creators?

**Comment by taion** [comment]
Right now the `TRANSITION` actions get swallowed. It won't mess with anything on the history side (or w/devtools) to let them through, but I don't know if that would lead to extra re-rendering or something, even if the store isn't actually updated at all by the reducers.

**Comment by jlongster** [comment]
Right, but it's a middleware that swallows it. If any middleware is run before it, it will see the the action. It's common in redux to specify the order you must list middleware, so I just want to make sure we say something about it if we need to. Like "make sure you put this before devtoolsMiddleware()"

**Comment by taion** [comment]
I thought the dev tools (1) weren't a middleware and (2) always go at the end anyway? But yeah, this should definitely go before the dev tools.

**Comment by jlongster** [comment]
Ah, it's not a middleware, it's a "store enhancer" like `applyMiddleware`. So we don't need to explain anything in the docs, it's already common practice that middleware gets executed first.

**Comment by jlongster** [comment]
I like it. Let's merge it and just see what else we need to fix up for a new version. (new docs obviously)

**Comment by timdorr** [comment]
If you need any help with this, I actually *do* use Redux and Router regularly. Though I haven't use a binding library (yet), I can look at anything from the perspective of The Redux Way™ if you need.

**Comment by taion** [comment]
I think the biggest open issue is action creator naming, per https://github.com/rackt/redux-simple-router/pull/141/files#r48519831 and follow-ups.

**Comment by timdorr** [comment]
An entirely unofficial, but popular, format for actions that many libraries use is a Flux Standard Action: https://github.com/acdlite/flux-standard-action

You might try putting information on the meta field. That is where I'm inclined to investigate. 

But let me check this out and integrate it into my code base to really get a feel for it. Is this pushed up to NPM?

**Comment by taion** [comment]
I don't mean the format (though we can namespace everything back under `payload` again if we want). I actually mean what to call the action creator functions. `push` and `replace` are too generic by a lot.

File: src/index.js (hunk 2 above)
Comment (timdorr) [inline src/index.js]: This could work:
```js
import { routeActions } from 'redux-simple-router'
routeActions.push('/foo')
```
It would also make using `bindActionCreators` easier because they're all under the same object namespace.

Comment (timdorr) [inline src/index.js]: Case in point:
```js
@connect(
  someSelector,
  dispatch => ({ router: bindActionCreators({ routeActions }, dispatch) })
)
class FooBar extends Component {
  handleClick = () => {
    this.props.router.push("/foo")
  }
}
```
I like that a lot.

Comment (taion) [inline src/index.js]: PR it then :p

Comment (timdorr) [inline src/index.js]: Done! #149

**Comment by neverfox** [comment]
Great work on all of this! Sounds like my `goBack` button will now work without having to be pressed twice. ;)

What's the currently release plan for all of the new goodies on master?

**Comment by gaearon** [comment]
Can somebody help me understand why `syncHistoryToStore` is necessary? I haven't seen methods on middleware before and it looks like boilerplate. Ideally passing the middleware should be enough for it to run. What am I missing?

**Comment by taion** [comment]
@gaearon It's for the dev tools. It's to enable the feature where, if the users manipulates the sequence of history-related actions, we sync the current history location to what's in the store.

**Comment by gaearon** [comment]
Yeah, but why is it separate? Is it incurring a production cost? Should it explicitly mention devtools? Right now the API makes it seem like it's necessary to call.

**Comment by taion** [comment]
@gaearon 

What do you mean "why is it separate"? My impression is that I can't do this within just the middleware itself, since the middleware runs before the dev tools.

Is there a more appropriate idiom here that would both let me transform the action before it hits the dev tools, and also subscribe to changes in the store?

One benefit, though, to setting this up this way, is that only the store subscriber cares about where the routing state ends up in the store - the middleware itself only works on the level of the actions.

**Comment by gaearon** [comment]
>What do you mean "why is it separate"? My impression is that I can't do this within just the middleware itself, since the middleware runs before the dev tools.

Not necessarily:

```
compose(
  applyMiddleware(a, b),
  DevTools.instrument(),
  applyMiddleware(c, d)
)
```

But unfortunately I don't understand the problem well enough to answer your question. If you have time, I'd appreciate a detailed writeup of what happens with and without that call, using some specific example (user went there, user pressed that, etc).

**Comment by taion** [comment]
I'd like to understand what the correct patterns here are. Right now, the code works like this:

1. User dispatches a `TRANSITION` action
2. The middleware listens to this `TRANSITION` action – it then calls e.g. `history.push` based on the contents of the `TRANSITION` action
3. `history` updates with the new location and calls its subscribers
  - **NOTE:** This is also the entry point for transitions triggered by e.g. `history.push` or a React Router `<Link>` directly, i.e. the ones that don't go through Redux actions
4. The `history` subscriber installed by the middleware dispatches an `UPDATE_LOCATION` action with the new location
5. The middleware in (2) doesn't do anything to this action
6. The dev tools observe this action and record it (as a staged action or something?)
7. Schematically, the relevant reducer handles this action to make the location available to store observers

Now the question is what happens when the developer cancels this action in the dev tools.

My understanding is that in most cases, cancelling an action should have no further side effects – it's sufficient for the dev tools to just update the store.

However, in some cases, when working with `history`, it might be desirable to actually also cause a side effect of transitioning to the previous location when cancelling `UPDATE_LOCATION` actions, e.g. from (4) above.

The way that `syncHistoryToStore` effects this (and only this) is to install a subscriber on the store. When it sees the store update with a location that wasn't the one most recently published by the `history` as in (3) above, it will call into the history to sync the location in history with the one in the store.

This might just be my lack of understanding, but it seemed like subscribing to the store for updates caused by the user e.g. cancelling actions in the dev tools was the best way to handle this, because at a glance it looked like the dev tools was publishing its calculated store state, rather than replaying some set of actions.

**Comment by gaearon** [comment]
If you want to subscribe to the store (and it seems that you do), the “Redux way” of doing that would be to turn this middleware into a store enhancer (sounds like Redux Router :-).

**Comment by taion** [comment]
Makes sense. The other two thoughts were:

1. This thing is a weird special-case side effect, because I only care about the terminal state the store ends up in; I don't want to e.g. replay all the `UPDATE_LOCATION` actions on the `history`
2. Maybe you don't always want to sync history to the store – the middleware works only on the level of the actions, whereas subscribing to the store seems to require me to make assumptions as to how the reducer is installed, which seems a bit more involved

**Comment by gaearon** [comment]
Can there be two different middlewares for that?

```js
applyMiddleware(reduxSimpleRouter),
DevTools.instrument(),
applyMiddleware(syncHistoryWithDevTools)
```

?

**Comment by taion** [comment]
Actually, looking at this a bit more, do you think it might be better to make this a dev tools monitor instead? `<SyncHistoryMonitor>` or something.

**Comment by gaearon** [comment]
FWIW I'd release a version that doesn't sync the URL back first. I'm not convinced it's that useful to perform side effect in this case. Then we can see if there's enough demand for it and consider different options.

**Comment by taion** [comment]
I'm totally on board with that. @tomatau mentioned it earlier in https://github.com/rackt/redux-simple-router/pull/141#issuecomment-167618037 as well, though in reference to an earlier, buggier implementation. It makes things a lot messier, and I have no idea what the workflow benefit actually is.

It's worth noting that this is exactly what you get right now if you just don't call `middleware.syncHistoryToStore`, though, and intentionally so, since this did seem very secondary to me.

I just didn't feel inclined to delete the whole "dev tools compatibility" section of the test suite.

**Comment by gaearon** [comment]
I think "compatibility" means "keeps working predictably regardless of what's in URL bar". As long as it doesn't crash with DevTools open and lets you go back in time, with the app updating visually to render previous routes, I think it's compatible. 

Now if the app doesn't update to render the previous routes, that would be a problem. I'm not sure what is considered the source of truth for the location React Router sees. Can we make the app update visually when you toggle history actions (and screw URL bar)?

**Comment by taion** [comment]
Not easily – you'd need something like https://github.com/rackt/redux-simple-router/issues/138, where Redux actually controls the `history` that the router uses as its source of truth for the current location, or at least a more restricted version that injects rewinded location states into the router.

I think that takes us back to redux-router-like levels of complexity.

But I think the complexity is just about equivalent with this implementation – you just call into `history.transitionTo` again, and you end up where you want to be with respect to the router.

Ultimately the issue here is that, since Redux isn't driving the router state, we need to do _something_ impure after dev tools actions to sync the router state back up to what's in the store. I think something like the current approach with an explicit "sync" makes the non-dev-tools code path a lot simpler and more predictable – we just need to be a bit careful with edge case handling on syncing the history after manipulating action history.

Whether that should be a store enhancer or whatever – that's just shuffling around a few lines of code; I think the basic logic right now is essentially correct.

**Comment by ezekielchentnik** [comment]
@gaearon I'm curious as to when it's appropriate to create a store enhancer. I understand the differences between middleware and store enhancers, but is there some guidance as to when it is a good or bad idea to create a store enhancer vs handling functionality externally?

**Comment by tomatau** [comment]
@taion & @gaearon 

Working with `syncHistoryToStore` is still not intuitive:
 * when rolling back, the app state rolls as expected but the path doesn't change (for me)
 * if actions are tied to routes, new actions can be dispatched when working with the dev tools - this doesn't align with the idea of "replaying state".

So, I still feel that `syncHistoryToStore` should be removed from `redux-simple-router`.  Both the ideas of making a new "SimpleRouterHistoryMonitor" or another middleware for this purpose seem sensible.

@taion - the internal complexity of redux-router was nowhere near as big a turn off as the complexity of initial setup for consumers. Redux-simple-router will never reach the same level of API complexity, so I think we shouldn't be putting too much weight on internal complexity when deciding an implementation.

Surely, the aim should be for the most well integrated solution, and so far, from the different approaches I've tested out, having redux drive the history updates using a history wrapper was the best for that... On the other hand, this middleware approach feels the most "reduxy" and least imposing. So for me, both of these approaches hold similar weight.

I'd quite like to see the "redux drive history" approach to be explored further, as it seems it was shot down prematurely due to worries of internal complexity, where really, it was doing quite good job of keeping state and history in sync. History methods just worked. DevTools just worked.

If now is the time to decide how this library will be implemented, it would be worth giving all the solutions a good kicking around and then we can all have more confidence that the best solution was chosen.

**Comment by jlongster** [comment]
@tomatau If rolling back actions does not change the URL, please give detailed steps to reproduce. When I last tried this, it worked for me. And the whole point of `transitionTo` inside the store subscriber was to change the URL without dispatching actions.

I think we definitely have room to play around with things, I'm more open to things like middleware/store enhancers now and we aren't near the complexity that redux router was. It had some good ideas though. I do think we should avoid wrapping history, etc. We can make everything work beautifully without too much work, but getting devtools to work is the most complicated, and from what I've heard redux-router didn't work great with it as well.

> FWIW I'd release a version that doesn't sync the URL back first. I'm not convinced it's that useful to perform side effect in this case. Then we can see if there's enough demand for it and consider different options.

If by syncing URL you mean to rerender the route when actions are rolled back, that's an often-requested feature and many of the bugs that come in are from people using devtools and expected it to "just work".

**Comment by mjrussell** [comment]
@tomatau @jlongster the path not changing is probably due to #164, I was playing with it a lot after spending some time upgrading/slimming down redux-router (following React-Router's v2) and running across similar history replace issues and was curious how redux-simple-router was implementing them.

It only happens when hitting the POP's (in this case resetting an action that directly follows a POP) because it cannot correctly replay the POP event when it is adding to the store.

**Comment by mjrussell** [comment]
I tried writing a breaking test for it, but trigger a POP event is actually pretty hard unless you have full control over the browser and the tests are currently designed to run without knowledge of the browser (which works great in every other case)

**Comment by taion** [comment]
@mjrussell 

Just use `createMemoryHistory` and call `goBack` or whatever. That said, I just don't know how to deal with this right now.

The problem with the "Redux drives history" approach is that, right now, `history` doesn't even drive history – it's just a proxy for the browser history API.

I don't think in a production use case, you want to deal with trying to completely hijack the browser history API. That API isn't designed for this, and it doesn't really play nicely with this sort of control.

I'm fine with complexity - we actually have a ton of complexity in the RR and history RCs for handling backward compat. What I'm not okay with is flakiness, and I think hijacking the browser history APIs is going to lead to unacceptable flakiness for production, user-facing pages.

That said, "Redux drives history" might be a decent approach for playing nicely with dev tools...

**Comment by tomatau** [comment]
> @tomatau If rolling back actions does not change the URL, please give detailed steps to reproduce. 

@jlongster I have a repo branch that you can see the behaviour in action. [here](https://github.com/tomatau/breko-hub/tree/rSrTest)

This is the syncHistoryToStore call on the client https://github.com/tomatau/breko-hub/blob/rSrTest/src/app/entry.js#L23

And here is where the middleware is added on the client  https://github.com/tomatau/breko-hub/blob/rSrTest/src/app/state/middleware.js#L17

It will sporadically update the path when routes are updated through the DevTools, but 9/10 times, the path doesn't update. Also, when the path does update via the DevTools, it's coming from a new action, which it shouldn't be.

This is on `"redux-simple-router": "^2.0.1"`

**Comment by taion** [comment]
Can you step through ~https://github.com/rackt/redux-simple-router/blob/master/src/index.js#L84-L102 to see what's happening? Unless you're completely wiping history (in which case we _need_ to push a new action to get the store into the correct state), you should hit the forward branch there, and as long as `syncing` is set, the history shouldn't update the store.

**Comment by jlongster** [comment]
Do you all mind looking through the docs here: https://github.com/rackt/redux-simple-router/blob/master/README.md particularly the API section? Is everything accurate?

I will push out 2.0.0 today or tomorrow.

**Comment by taion** [comment]
- The middleware has an `unsubscribe` method that unsubscribes everything
- The reducer is only necessary if you're using `syncHistoryToStore` (or if you want to, you know, actually observe the location); the middleware itself will work without it otherwise
- `goForward` and `goBack` have no args, while `go` just takes a `n` (signed int)

**Comment by timdorr** [comment]
>  -The middleware has an unsubscribe method that unsubscribes everything

Fixed!

> - The reducer is only necessary if you're using `syncHistoryToStore` (or if you want to, you know, actually observe the location); the middleware itself will work without it otherwise

Correct, but I wouldn't want to lead beginners down the path of thinking it's optional and then having them be confused when `state.routing` is empty.

> - `goForward` and `goBack` have no args, while `go` just takes a `n` (signed int)

Fixed!

**Comment by billyjanitsch** [comment]
Sorry that I'm commenting after v2 has shipped, but for what it's worth, I think this new API has bad implications for separation of concerns, and is a lot more complicated to setup.

The example given in the [tutorial](https://github.com/rackt/redux-simple-router/blob/master/README.md#tutorial) looks reasonable enough until you realize that a typical Redux setup will be doing much more than the following two lines:

```js
const createStoreWithMiddleware = applyMiddleware(reduxRouterMiddleware)(createStore)
const store = createStoreWithMiddleware(reducer)
```

Think additional middleware, store enhancers, reducer updating, etc. A [very](https://github.com/rackt/redux/blob/master/examples/real-world/store/configureStore.dev.js) [common](https://github.com/rackt/redux/blob/master/examples/async/store/configureStore.js) [pattern](https://github.com/yelouafi/redux-saga/blob/master/examples/real-world/store/configureStore.dev.js) is to abstract away this setup into a separate function which returns an enhanced `createStore` function, parameterized according to the [Redux spec](http://redux.js.org/docs/api/createStore.html). This is a really clean abstraction.

However, using redux-simple-router@2, `createStore` now requires access to the `history` singleton to pass to `syncHistory`. `createStore` can't initialize it because it needs to be passed around elsewhere (e.g. to react-router), but you can't further parameterize `createStore` without breaking its natural signature (preventing further store enhancement). It also must now hook onto its own store creation so that `listenForReplays` can be called on the store once it's created.

At that point, if you're already entangling all of this history logic in your store creation -- left as boilerplate for the user -- why isn't redux-simple-router just a store enhancer? This would eliminate the need for `listenForReplays` and allow `createStore` to be enhanced by redux-simple-router at the top level. Your history logic stays separate, there's less boilerplace, you don't need to break `createStore`, and you get the same behavior.

**Comment by timdorr** [comment]
react-router embraces singleton histories for version 2.0.0 (coming out soon!): https://github.com/rackt/react-router/blob/master/upgrade-guides/v2.0.0.md#history-singletons-provided

You can similarly define your history in a singleton manner, making passing it around a lot easier and not requiring you to include it as an argument to your `createStore` wrapper.

**Comment by mjrussell** [comment]
@billyjanitsch I agree that this new structure does typically require you pass the history singleton both to your createStore and your Router, but how would a store enhancer in this case improve that? 

As I see it the store enhancer would still need the history access and therefore you just moved the call from inside applyMiddleware to the compose but it still lies in the same code. Unless of course you pass down the store enhancer to this configureStore function but thats not really any better.

Maybe some psuedo js of how it would be used would make it more clear, perhaps Im just missing something.

(just saw @timdorr comment but thought I'd still ask the question)

**Comment by billyjanitsch** [comment]
@timdorr

> react-router embraces singleton histories for version 2.0.0 (coming out soon!)

Ah, thanks for the link, I hadn't seen this! That's a huge improvement, and solves most of my wall of text above aside from `listenForReplays`. :)

@mjrussell 

Sorry, I was unclear. The idea is that your store enhancers should work separately from each other so that you can (conditionally) compose them however you'd like. When `listenForReplays` (which is essentially a user-end store enhancer) depends on some parameter of `applyMiddleware` -- even with access to the `history` singleton -- it breaks this composability.

Perhaps I want to abstract my middleware:
```js
import {createStore, applyMiddleware} from 'redux'
import {syncHistory} from 'redux-simple-router'
import {browserHistory} from 'redux-router' // thanks @timdorr :)
import middleware from './myMiddleware'

const enhanceStore = compose(
  syncHistory(browserHistory),
  applyMiddleware(...middleware)
)

export default enhanceStore(createStore)
```

Or abstract most of my store enhancers:
```js
import createStore from './myCreateStore' // already enhanced by me
import {createHistory} from 'history' // stuck on react-router v1 :(
import myReducer from './myReducer'

const history = createHistory()
const finalCreateStore = syncHistory(history)(createStore)
const store = finalCreateStore(reducer)

render(<App store={store} history={history} />, container)
```

It's not any particular case; it's the general principle of keeping middleware and store enhancers abstractable. I believe that store enhancers were in fact designed to keep middleware from having to do this sort of thing.

**Comment by mjrussell** [comment]
@billyjanitsch Thanks I see where you are going, especially if you split your enhancer into a separate file then you do benefit from some decoupling in that case because you can do the sync to store call inside the enhancer.

**Comment by taion** [comment]
The main constraints here are:

1) The basic middleware is just like any other middleware (other than needing the `history`, which it can get from a singleton exported by some module), and is very simple
2) Without a special history per #138, you need something else after the dev tools to pick up on stuff devtools does
3) Ideally, the store enhancers from (2) and (1) can communicate to each other, to prevent (1) from generating new Redux actions when restoring history location, unless we go with #138 

I'd largely be happy with whatever – the main goal here was just to move to an implementation in (1) that naturally excludes the possibility of updating history twice, through the previous syncing method.

**Comment by taion** [comment]
And I'd add that you'd want to probably completely discard the replay logic when not in development.

**Comment by tomatau** [comment]
@taion, apologies for the delay.

> Can you step through ~https://github.com/rackt/redux-simple-router/blob/master/src/index.js#L84-L102 to see what's happening? Unless you're completely wiping history (in which case we need to push a new action to get the store into the correct state), you should hit the forward branch there, and as long as syncing is set, the history shouldn't update the store.

A POP is being passed to `history.transitionTo` with the correct previous path etc... This seems to just get swallowed. After which, history.listen gets invoked correctly, and the store is returning the correct location.key -- but the path __hasn't__ updated in the browser still.

NB: It's a little bit painful to debug with the DevTools or whatever monitor, repeatedly emitting store updates.

**Comment by taion** [comment]
@tomatau I believe that's https://github.com/rackt/redux-simple-router/issues/164, then? Where do you see the additional store updates come from?

**Comment by tomatau** [comment]
Looks very related to #164 

When I say `store updates` -- I'm talking about from the DevTools or LogMonitor itself, it seems to fire some sort of tick invoking the store.subscribe handler. Only a pain point for triggering the debugger.

**Comment by pdf** [comment]
After this refactor, how are we supposed to access the router state in an immutable store?

**Comment by mjrussell** [comment]
@pdf maybe you are looking for this? - https://github.com/rackt/react-router-redux/pull/218

**Comment by pdf** [comment]
Indeed, thanks @mjrussell
